core/dbt/clients/_jinja_blocks.py

```diff
@@ -417,156 +246,122 @@ def next_block(self, block_handler, include_toplevel=False):
-        {% macro my_macro(arg="x" + ("}% {# {% endmacro %}" * 2)) %}
+class BlockIterator(object):
+    def __init__(self, data):
+        self.tag_parser = TagIterator(data)
+        self.current = None
+        self.stack = []
+        self.last_position = 0
+
+    @property
+    def current_end(self):
+        if self.current is None:
+            return 0
+        else:
+            return self.current.end
 
-        Which makes you a jerk, but is valid jinja.
-        """
-        # we are currently after the first parenthesis (+ any whitespace) after
-        # the macro args started. You can either have the close paren, or a
-        # name.
-        while self._parenthesis_stack:
-            match = self._expect_match('macro arguments',
-                                       MACRO_ARGS_END_PATTERN,
-                                       MACRO_ARG_PATTERN)
-            self.advance(match.end())
-            matchgroups = match.groupdict()
-            if matchgroups.get('macro_end') is not None:
-                self._parenthesis_stack.pop()
-            # we got an argument. let's see what it has
-            elif matchgroups.get('value') is not None:
-                # we have to process a single macro argument. This mutates
-                # the parenthesis stack! If it finds a comma, it will continue
-                # the loop.
-                self._process_macro_default_arg()
-            elif matchgroups.get('more_args') is not None:
-                continue
-            else:
-                raise dbt.exceptions.InternalException(
-                    'unhandled regex in _process_macro_args(), no match: {}'
-                    .format(matchgroups)
+    @property
+    def data(self):
+        return self.tag_parser.data
+
+    def is_current_end(self, tag):
+        return (
+            tag.block_type_name.startswith('end') and
+            self.current is not None and
+            tag.block_type_name[3:] == self.current.block_type_name
+        )
+
+    def find_blocks(self, allowed=None, collect_raw_data=True):
+        """Find all top-level blocks in the data."""
+        if allowed is None:
+            allowed = {'snapshot', 'macro', 'materialization', 'docs'}
+
+        for tag in self.tag_parser.find_tags():
+            if tag.block_type_name in _CONTROL_FLOW_TAGS:
+                self.stack.append(tag.block_type_name)
+            elif tag.block_type_name in _CONTROL_FLOW_END_TAGS:
+                found = None
+                if self.stack:
+                    found = self.stack.pop()
+                expected = None
+                if found:
+                    expected = _CONTROL_FLOW_TAGS[found]
+                if expected != tag.block_type_name:
+                    dbt.exceptions.raise_compiler_error((
+                        'Got unexpected control flow end tag, got {} but '
+                        'expected {} next (@ {})'
+                    ).format(tag.block_type_name, expected, tag.start))
```

Comment:
i think if you hit a control flow end tag first, e.g. a file containing only:

```
{% endif %}
```

the error message will be:

`Got unexpected control flow end tag, got endif but expected None next (@ {% endif %})`

i think you need to handle the case where both found and expected are None before you get here